### ISSUE TITLE
Newton Method one dimensional

### ISSUE THREAD

**Issue by sdrap** [argmin-rs/argmin#345]
I am totally new to this library and ran a simple one dimensional test following the newton.rs example.

```rust
use argmin::core::observers::{ObserverMode, SlogLogger};
use argmin::core::{CostFunction, Error, Executor, Gradient, Hessian};
use argmin::solver::linesearch::MoreThuenteLineSearch;
use argmin::solver::newton::Newton;

struct MyFunction {
    a: f64,
    b: f64,
}

impl Gradient for MyFunction {
    type Param = f64;
    type Gradient = f64;

    fn gradient(&self, p: &Self::Param) -> Result<Self::Gradient, Error> {
        Ok(2.0 * *p * self.a + self.b)
    }
}

impl Hessian for MyFunction {
    type Param = f64;
    type Hessian = f64;

    fn hessian(&self, _p: &Self::Param) -> Result<Self::Hessian, Error> {
        Ok(self.a * 2.0)
    }
}

fn run() -> Result<(), Error> {
    // Define cost function
    let cost = MyFunction {
        a: 1.0,
        b: -2.0,
    };

    // Define initial parameter vector
    // let init_param: Array1<f64> = Array1::from(vec![1.2, 1.2]);
    let init_param: f64 = 1.5;

    // set up line search
    // Set up solver
    let solver = Newton::new();

    // Run solver
    let res = Executor::new(cost, solver)
        .configure(|state| state.param(init_param).max_iters(100))
        .add_observer(SlogLogger::term(), ObserverMode::Always)
        .run()?;

    // Wait a second (lets the logger flush everything before printing again)
    std::thread::sleep(std::time::Duration::from_secs(1));

    // Print result
    println!("This is the result {res}");
    Ok(())
}

fn main() {
    if let Err(ref e) = run() {
        println!("There is an error {e}");
        std::process::exit(1);
    }
}
```

I run into the error `f64: argmin_math::ArgminInv<f64> is not satisfied`.
I suppose that `ArgminInv` is expecting some kind of matrix type hessian and try to perform an inverse but in one dimension that's just a plain division. 

Is there a simple way to apply the newton method to find the root of the gradient in one dimension? In other terms use newton method as root finding even for one dimensional problems? (brentq is nice but sometimes not the best in some specific situations in particular for 1d solvers).

Many thanks in any cases, I am really looking forward to a robust and efficient optimization library.

**Comment by stefan-k** [comment]
Hi, 

thanks for pointing this out, there is definitely the `ArgminInv` impl missing for `f32` and `f64` (and potentially other traits too). Unfortunately, for now you have to resort to using a `Vec<f64>` with only a single element. 

I will open another issue today to keep track of this. 

Apologies for the inconvenience!

**Comment by sdrap** [comment]
No problem, I was looking a bit deeper in the internal and indeed, this `ArgminInv` makes a call to `ndarray-linalg inv` that computes the inverse of a matrix.
Maybe for one dimensional optimization problems, when inverse is needed, calling matrix inverse is probably not the most efficient method. It would require however to distinguish between 1 dim and 2+ dim solvers.
Maybe the easiest way is to create a separate 1d Newton method. I will have a look since I am quite interested into that one.

**Comment by stefan-k** [comment]
> It would require however to distinguish between 1 dim and 2+ dim solvers.
> Maybe the easiest way is to create a separate 1d Newton method.

I had a quick look and my take is that implementing `ArgminInv` for `f64` and `f32` as a plain division should be sufficient. Or am I missing something?

> I will have a look since I am quite interested into that one. 

That would be great! :)

**Comment by stefan-k** [comment]
Fixed by #346
